synapse/handlers/space_summary.py

```diff
@@ -332,45 +397,88 @@ async def _summarize_remote_room(
             or ev.event_type == EventTypes.SpaceChild
         )
 
-    async def _is_room_accessible(self, room_id: str, requester: Optional[str]) -> bool:
+    async def _is_room_accessible(
+        self, room_id: str, requester: Optional[str], origin: Optional[str]
+    ) -> bool:
         """
         Calculate whether the room should be shown in the spaces summary.
 
         It should be included if:
 
         * The requester is joined or invited to the room.
+        * The requester can join without an invite (per MSC3083).
+        * The origin server has any user that is joined or invited to the room.
         * The history visibility is set to world readable.
 
         Args:
             room_id: The room ID to summarize.
             requester:
                 The user requesting the summary, if it is a local request. None
                 if this is a federation request.
+            origin:
+                The server requesting the summary, if it is a federation request.
+                None if this is a local request.
 
         Returns:
              True if the room should be included in the spaces summary.
         """
+        state_ids = await self._store.get_current_state_ids(room_id)
+
+        # If there's no state for the room, it isn't known.
+        if not state_ids:
+            logger.info("room %s is unknown, omitting from summary", room_id)
+            return False
+
+        room_version = await self._store.get_room_version(room_id)
 
         # if we have an authenticated requesting user, first check if they are able to view
         # stripped state in the room.
         if requester:
-            try:
-                await self._auth.check_user_in_room(room_id, requester)
+            member_event_id = state_ids.get((EventTypes.Member, requester), None)
+
+            # If they're in the room they can see info on it.
+            if member_event_id:
+                member_event = await self._store.get_event(member_event_id)
+                if member_event.membership in (Membership.JOIN, Membership.INVITE):
+                    return True
+
+            # Otherwise, if they should be allowed access via membership in a space.
+            # Get the spaces which allow access to this room, if applicable.
+            allowed_spaces = await self._event_auth_handler.get_spaces_that_allow_join(
+                state_ids, room_version
+            )
+            if allowed_spaces and await self._event_auth_handler.is_user_in_rooms(
+                allowed_spaces, requester
+            ):
+                return True
```

Comment:
This is pretty much the same as `check_restricted_join_rules`? I wonder if should use that or add a helper function for it?

Reply:
We could add a helper (`can_join_via_restricted_join_rules`) which returns a boolean which is called here and `check_restricted_join_rules`?

Reply:
I attempted this but it seemed more confusing in a way. I ended up doing a call `check_restricted_join_rules` and catching the `AuthError`.